lsp/state/workspace.cpp

```diff
@@ -0,0 +1,290 @@
+    // NOTE - assumes we can only find 1 function per line
+    if (info_matches.size() == 2) {
+      auto match = info_matches[1];
+      lg::debug("Found info-level diagnostic - {}", match.str());
+      LSPSpec::Diagnostic new_diag;
+      new_diag.m_severity = LSPSpec::DiagnosticSeverity::Information;
+      new_diag.m_message = match.str();
+      new_diag.m_range = diag_range;
+      new_diag.m_source = "OpenGOAL LSP";
+      m_diagnostics.push_back(new_diag);
+      return;
+    }
+  }
+  // Check for a warn level warnings
+  if (std::regex_search(line, warn_matches, warn_regex)) {
+    // NOTE - assumes we can only find 1 function per line
+    if (warn_matches.size() == 2) {
+      auto match = warn_matches[1];
+      lg::debug("Found warn-level diagnostic - {}", match.str());
+      LSPSpec::Diagnostic new_diag;
+      new_diag.m_severity = LSPSpec::DiagnosticSeverity::Error;
+      new_diag.m_message = match.str();
+      new_diag.m_range = diag_range;
+      new_diag.m_source = "OpenGOAL LSP";
+      m_diagnostics.push_back(new_diag);
+      return;
+    }
+  }
+}
+
+std::optional<std::string> WorkspaceIRFile::get_mips_instruction_at_position(
+    const LSPSpec::Position position) {
+  // Split the line on typical word boundaries
+  std::string line = m_lines.at(position.m_line);
+  std::smatch matches;
+  std::regex regex("[\\w\\.]+");
+
+  if (std::regex_search(line, matches, regex)) {
+    auto match = matches[0];
+    lg::info("hover first match - {}", match.str());
+    auto match_start = matches.position(0);
+    auto match_end = match_start + match.length();
+    if (position.m_character >= match_start && position.m_character <= match_end) {
+      return match;
+    }
+  }
+
+  return {};
+}
+
+std::optional<std::string> WorkspaceIRFile::get_symbol_at_position(
+    const LSPSpec::Position position) {
+  // Split the line on typical word boundaries
+  std::string line = m_lines.at(position.m_line);
+  lg::debug("symbol checking - '{}'", line);
+  std::smatch matches;
+  std::regex regex("[\\w\\.\\-_!<>*]+");
+  std::regex_token_iterator<std::string::iterator> rend;
+
+  std::regex_token_iterator<std::string::iterator> match(line.begin(), line.end(), regex);
+  while (match != rend) {
+    lg::debug("match - '{}'", match->str());
+    auto match_start = std::distance(line.begin(), match->first);
+    auto match_end = match_start + match->length();
+    if (position.m_character >= match_start && position.m_character <= match_end) {
+      lg::debug("returning");
+      return match->str();
+    }
+    match++;
+  }
+
+  return {};
+}
+
+void WorkspaceAllTypesFile::parse_type_system() {
+  lg::debug("DTS Loading - '{}'", m_file_path.string());
+  m_dts.parse_type_defs({m_file_path.string()});
+  lg::debug("DTS Loaded At - '{}'", m_file_path.string());
+}
+
+void WorkspaceAllTypesFile::update_type_system() {
+  // TODO - do i actually need to re-create it, or is there a way to "clear" it?
```

Comment:
recreating it is fine - I prefer to do things that way because there's no need to write a separate `reset` method, and I think objects that are easy to reconstruct like this are a good pattern.